### PR TITLE
Refactor estimation

### PR DESCRIPTION
**Pull request by kisieland**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

Introduces optimization to the binpacking estimator, which changes the time complexity from O(pods * nodes) to O(pods) in cases of homogeneous pods.

The PR also introduces a benchmark that shows 5x improvement:
The previous version took ~25s:
```
goos: linux
goarch: amd64
pkg: k8s.io/autoscaler/cluster-autoscaler/estimator
cpu: AMD EPYC 7B12
BenchmarkBinpackingEstimate-48    	       1	25418848308 ns/op	15018690016 B/op	144793817 allocs/op
PASS
ok  	k8s.io/autoscaler/cluster-autoscaler/estimator	25.488s
```
The new version takes less than 5s:
```
goos: linux
goarch: amd64
pkg: k8s.io/autoscaler/cluster-autoscaler/estimator
cpu: AMD EPYC 7B12
BenchmarkBinpackingEstimate-48    	       1	4805105872 ns/op	2023739088 B/op	 3574976 allocs/op
PASS
ok  	k8s.io/autoscaler/cluster-autoscaler/estimator	4.869s
```

#### Special notes for your reviewer:
Split the PR into two commits:
First one introduces a required pod-equivalence abstraction
Second one uses this new fact in the estimation

For now I kept the logic in one function in the estimation to make clear what changes were made, but I am happy to split it into further functions.

#### Does this PR introduce a user-facing change?
No

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
NA

### CODE REVIEW

File: cluster-autoscaler/core/scaleup/orchestrator/orchestrator.go [hunk 1]
```diff
@@ -683,22 +685,26 @@ func (o *ScaleUpOrchestrator) ComputeSimilarNodeGroups(
...
 
 	return validSimilarNodeGroups
 }
 
-func matchingSchedulablePods(groupSchedulablePods []*apiv1.Pod, similarSchedulablePods []*apiv1.Pod) bool {
+func matchingSchedulablePods(podGroups []estimator.PodEquivalenceGroup, similarPodGroups []estimator.PodEquivalenceGroup) bool {
 	schedulablePods := make(map[*apiv1.Pod]bool)
-	for _, pod := range similarSchedulablePods {
-		schedulablePods[pod] = true
+	for _, podGroup := range similarPodGroups {
```
Comment (x13n) [inline cluster-autoscaler/core/scaleup/orchestrator/orchestrator.go:698]: Can we just compare groups not the pods in them? The groups should be immutable, so iterating over all pods seems wasteful.

File: cluster-autoscaler/estimator/binpacking_estimator.go [hunk 2]
```diff
@@ -89,14 +89,19 @@ func (e *BinpackingNodeEstimator) Estimate(
 	scheduledPods := []*apiv1.Pod{}
 	lastNodeName := ""
 
-	for _, pod := range pods {
-		found := false
+	for _, podsEquivalenceGroup := range podsEquivalenceGroups {
```
Comment (x13n) [inline cluster-autoscaler/estimator/binpacking_estimator.go:92]: Can you split this into smaller functions? It's pretty long right now, long nested loops are hard to read and there's a bit of code duplication that perhaps could be avoided.

File: cluster-autoscaler/estimator/binpacking_estimator.go [hunk 3]
```diff
@@ -105,47 +110,66 @@ func (e *BinpackingNodeEstimator) Estimate(
...
-			// on a new node either. There is no point adding more nodes to snapshot in such case, especially because of
-			// performance cost each extra node adds to future FitsAnyNodeMatching calls.
-			if lastNodeName != "" && !newNodesWithPods[lastNodeName] {
-				continue
+		for ; index < len(podsEquivalenceGroup.Pods); index++ {
+			pod := podsEquivalenceGroup.Pods[index]
+			found := false
+
+			// Check schedulability on only newly created node
+			nodeName, err := e.predicateChecker.FitsAnyNodeMatching(e.clusterSnapshot, pod, func(nodeInfo *schedulerframework.NodeInfo) bool {
```
Comment (x13n) [inline cluster-autoscaler/estimator/binpacking_estimator.go:118]: Since we only want to check one node here, perhaps it would suffice to call `CheckPredicates` instead? That'll avoid listing all nodes from cluster snapshot just to pick one of them later on. Also, no point in running this check when `lastNodeName` is empty I think?

File: cluster-autoscaler/estimator/binpacking_estimator.go [hunk 4]
```diff
@@ -105,47 +110,66 @@ func (e *BinpackingNodeEstimator) Estimate(
...
+				newNodeNameIndex++
+				newNodeNames[newNodeName] = true
+				lastNodeName = newNodeName
+
+				// And try to schedule pod to it.
+				// Note that this may still fail (ex. if topology spreading with zonal topologyKey is used);
+				// in this case we can't help the pending pod. We keep the node in clusterSnapshot to avoid
+				// adding and removing node to snapshot for each such pod.
+				if err := e.predicateChecker.CheckPredicates(e.clusterSnapshot, pod, newNodeName); err != nil {
+					continue
```
Comment (x13n) [inline cluster-autoscaler/estimator/binpacking_estimator.go:164]: Shouldn't we skip the entire group in this case? If one pod failed, what's the point of trying another identical pod?

File: cluster-autoscaler/estimator/binpacking_estimator_test.go [hunk 5]
```diff
@@ -79,10 +79,12 @@ func makePods(cpuPerPod int64, memoryPerPod int64, hostport int32, maxSkew int32
...
 		pods = append(pods, pod)
 	}
-	return pods
+	return PodEquivalenceGroup{
+		Pods: pods,
+	}
 }
 
-func makeNode(cpu int64, mem int64, name string, zone string) *apiv1.Node {
+func makeNode(cpu, mem, podCount int64, name string, zone string) *apiv1.Node {
```
Comment (x13n) [inline cluster-autoscaler/estimator/binpacking_estimator_test.go:87]: Similar to the above. Could this use `BuildTestNode` instead? https://github.com/kubernetes/autoscaler/blob/7184d23e62367c57b2f0d66203a5d7c7a7eb661e/cluster-autoscaler/utils/test/test_utils.go#L215

Missing params can be added via option pattern. Right now it is hard to tell what `makeNode(1, 2, 3, "4", "5")` is doing without referring back here.

File: cluster-autoscaler/estimator/binpacking_estimator_test.go [hunk 6]
```diff
@@ -32,7 +32,7 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-func makePods(cpuPerPod int64, memoryPerPod int64, hostport int32, maxSkew int32, topologySpreadingKey string, podCount int) []*apiv1.Pod {
+func makePodEquivalenceGroup(cpuPerPod int64, memoryPerPod int64, hostport int32, maxSkew int32, topologySpreadingKey string, podCount int) PodEquivalenceGroup {
```
Comment (x13n) [inline cluster-autoscaler/estimator/binpacking_estimator_test.go:35]: I know you're just modifying an existing function, but could this just accept a pod and count to be composed with `BuildTestPod`? https://github.com/kubernetes/autoscaler/blob/7184d23e62367c57b2f0d66203a5d7c7a7eb661e/cluster-autoscaler/utils/test/test_utils.go#L36

For instance: `makePodEquivalenceGroup(BuildTestPod(...), 10)`? Missing params can be added via option pattern.

**Review by x13n — CHANGES_REQUESTED** [review]

File: cluster-autoscaler/core/scaleup/orchestrator/orchestrator.go (hunk 1 above)
Comment (kisieland) [reply cluster-autoscaler/core/scaleup/orchestrator/orchestrator.go:698]: Made the function take a look only on the sample pod (i.e. the first pod in the slice)

File: cluster-autoscaler/estimator/binpacking_estimator_test.go (hunk 6 above)
Comment (kisieland) [reply cluster-autoscaler/estimator/binpacking_estimator_test.go:35]: kept the `BuildTestPod` hidden as a call inside in `makePodEquivalenceGroup`
But this simplifies the code quite.

File: cluster-autoscaler/estimator/binpacking_estimator.go (hunk 3 above)
Comment (kisieland) [reply cluster-autoscaler/estimator/binpacking_estimator.go:118]: Make sense, rewrote.

File: cluster-autoscaler/estimator/binpacking_estimator.go (hunk 4 above)
Comment (kisieland) [reply cluster-autoscaler/estimator/binpacking_estimator.go:164]: Done.

File: cluster-autoscaler/estimator/binpacking_estimator_test.go (hunk 5 above)
Comment (kisieland) [reply cluster-autoscaler/estimator/binpacking_estimator_test.go:87]: Done

File: cluster-autoscaler/estimator/binpacking_estimator.go [hunk 7]
```diff
@@ -68,49 +78,98 @@ func NewBinpackingNodeEstimator(
...
-	newNodeNameIndex := 0
 	scheduledPods := []*apiv1.Pod{}
-	lastNodeName := ""
+	for _, podsEquivalenceGroup := range podsEquivalenceGroups {
+		var err error
+		var remainingPods []*apiv1.Pod
+
+		scheduledPods, remainingPods, err = e.tryToScheduleOnExistingNodes(scheduledPods, podsEquivalenceGroup.Pods)
+		if err != nil {
+			klog.Errorf("%v", err)
```
Comment (x13n) [inline cluster-autoscaler/estimator/binpacking_estimator.go:104]: nit: This is equivalent to `klog.Errorf(err.Error())`, just with unnecessary formatting. Same below.

File: cluster-autoscaler/estimator/binpacking_estimator.go [hunk 8]
```diff
@@ -68,49 +78,98 @@ func NewBinpackingNodeEstimator(
...
-		if err == nil {
-			found = true
-			if err := e.clusterSnapshot.AddPod(pod, nodeName); err != nil {
-				klog.Errorf("Error adding pod %v.%v to node %v in ClusterSnapshot; %v", pod.Namespace, pod.Name, nodeName, err)
-				return 0, nil
+		if err != nil {
+			break
+		}
+
+		if err := e.clusterSnapshot.AddPod(pod, nodeName); err != nil {
```
Comment (x13n) [inline cluster-autoscaler/estimator/binpacking_estimator.go:137]: This pattern repeats in 3 places: add pod to cluster snapshot, append to `scheduledPods`, update `newNodesWithPods`. Perhaps worth extracting to a function?

File: cluster-autoscaler/estimator/binpacking_estimator.go [hunk 9]
```diff
@@ -124,49 +183,45 @@ func (e *BinpackingNodeEstimator) Estimate(
...
-			if err != nil {
-				klog.Errorf("Error while adding new node for template to ClusterSnapshot; %v", err)
-				return 0, nil
+			if err := e.addNewNodeToSnapshot(nodeTemplate); err != nil {
+				return nil, fmt.Errorf("Error while adding new node for template to ClusterSnapshot; %w", err)
 			}
-			newNodeNameIndex++
-			newNodeNames[newNodeName] = true
-			lastNodeName = newNodeName
+			e.newNodeNames[e.lastNodeName] = true
```
Comment (x13n) [inline cluster-autoscaler/estimator/binpacking_estimator.go:189]: nit: make it a part of `addNewNodeToSnapshot`?

File: cluster-autoscaler/estimator/binpacking_estimator_test.go (hunk 6 above)
Comment (x13n) [reply cluster-autoscaler/estimator/binpacking_estimator_test.go:35]: I think with 6 unnamed variables, mostly of the same type, it is hard to read test cases without referring back here. Hence my proposal to explicitly call `BuildTestPod` there - the options are named, is it is clear what they do.

File: cluster-autoscaler/estimator/binpacking_estimator_test.go (hunk 5 above)
Comment (x13n) [reply cluster-autoscaler/estimator/binpacking_estimator_test.go:57]: Is it? I still see `makeNode` not `BuildTestNode` here.

**Review by x13n — CHANGES_REQUESTED** [review]

File: cluster-autoscaler/estimator/binpacking_estimator.go [hunk 10]
```diff
@@ -36,6 +36,12 @@ type BinpackingNodeEstimator struct {
 	podOrderer             EstimationPodOrderer
 	context                EstimationContext
 	estimationAnalyserFunc EstimationAnalyserFunc // optional
+
+	// Helper variables to avoid coping them in each function
+	newNodeNameIndex int
```
[the remark this answers is not in the document]
Comment (kisieland) [reply cluster-autoscaler/estimator/binpacking_estimator.go:41]: SG, done

File: cluster-autoscaler/core/scaleup/orchestrator/orchestrator.go (hunk 1 above)
Comment (kisieland) [reply cluster-autoscaler/core/scaleup/orchestrator/orchestrator.go:698]: Added, though there were only few calls added to this function.

File: cluster-autoscaler/estimator/binpacking_estimator_test.go (hunk 6 above)
Comment (kisieland) [reply cluster-autoscaler/estimator/binpacking_estimator_test.go:35]: Updated

File: cluster-autoscaler/estimator/binpacking_estimator_test.go (hunk 5 above)
Comment (kisieland) [reply cluster-autoscaler/estimator/binpacking_estimator_test.go:57]: Updated to use `BuildTestNode` directly.
